File: agent/scheduler.py
from datetime import datetime, time
from typing import Any


def parse_time(value: str) -> time:
    parts = value.split(":")
    return time(int(parts[0]), int(parts[1]))


def is_within_schedule(config: dict[str, Any], now: datetime | None = None) -> bool:
    now = now or datetime.now()
    schedule = config.get("schedule", {})
    enabled_days = schedule.get("enabled_days", "0,1,2,3,4,5,6")
    allowed = {int(d.strip()) for d in enabled_days.split(",") if d.strip().isdigit()}
    if now.weekday() not in allowed:
        return False

    start = parse_time(str(schedule.get("start_time", "07:00:00"))[:8])
    end = parse_time(str(schedule.get("end_time", "18:00:00"))[:8])
    current = now.time()
    if start <= end:
        return start <= current <= end
    return current >= start or current <= end


def interval_seconds(config: dict[str, Any]) -> int:
    schedule = config.get("schedule", {})
    value = int(schedule.get("interval_value", 5))
    unit = schedule.get("interval_unit", "minutes")
    if unit == "seconds":
        return max(1, value)
    return max(1, value * 60)

File: firmware/esp32-cam-eth/main/scheduler.c
#include "scheduler.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static int parse_time_seconds(const char *value)
{
    int h = 0;
    int m = 0;
    int s = 0;
    sscanf(value, "%d:%d:%d", &h, &m, &s);
    return h * 3600 + m * 60 + s;
}

bool scheduler_is_within_schedule(const schedule_config_t *schedule, time_t now)
{
    struct tm tm_now;
    localtime_r(&now, &tm_now);

    // tm_wday de C usa domingo=0..sábado=6; el backend (y agent/scheduler.py,
    // basado en datetime.weekday() de Python) usa lunes=0..domingo=6.
    int python_weekday = (tm_now.tm_wday + 6) % 7;

    bool day_enabled = false;
    char days_copy[32];
    strlcpy(days_copy, schedule->enabled_days[0] ? schedule->enabled_days : "0,1,2,3,4,5,6",
             sizeof(days_copy));
    char *saveptr = NULL;
    for (char *tok = strtok_r(days_copy, ",", &saveptr); tok; tok = strtok_r(NULL, ",", &saveptr)) {
        if (atoi(tok) == python_weekday) {
            day_enabled = true;
            break;
        }
    }
    if (!day_enabled) {
        return false;
    }

    int start = parse_time_seconds(schedule->start_time[0] ? schedule->start_time : "07:00:00");
    int end = parse_time_seconds(schedule->end_time[0] ? schedule->end_time : "18:00:00");
    int current = tm_now.tm_hour * 3600 + tm_now.tm_min * 60 + tm_now.tm_sec;

    if (start <= end) {
        return current >= start && current <= end;
    }
    return current >= start || current <= end;
}

int scheduler_interval_seconds(const schedule_config_t *schedule)
{
    int value = schedule->interval_value > 0 ? schedule->interval_value : 5;
    if (strcmp(schedule->interval_unit, "seconds") == 0) {
        return value < 1 ? 1 : value;
    }
    int seconds = value * 60;
    return seconds < 1 ? 1 : seconds;
}
